### ISSUE TITLE
Unable to use resources.get_controller() function

### ISSUE THREAD

**Issue by Tama-sha** [saezlab/pypath#226]
Steps used:
from pypath import resources
rc = resources.get_controller()
rc

Error:
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 40814: character maps to <undefined>

Complete snapshot:
![image](https://user-images.githubusercontent.com/106139199/224232097-bd02268c-f332-4b71-b33b-10dc88ba3593.png)

Additional query: 
I wanted to create a network from resources available for commercial use but i am not able to obtain the correct function either

**Comment by deeenes** [comment]
Hello @Tama-sha, The reason of this error is that your operating system's default character encoding is code page 1252, while the `resources.json` file distributed with `pypath` is UTF-8 encoded, as it is more or less a standard practice today. The solution is to run your Python interpreter with UTF-8 default encoding.

**Comment by deeenes** [comment]
About your other question, I checked the current state of license parameters in pypath, and I filled a few gaps to make it better usable in the database build process. Due to these recent developments, you'll need to update pypath from the git master branch (version should be `>=0.14.38`). I also [added a little section](https://pypath.omnipathdb.org/notebooks/manual.html#Example:-build-a-network-for-commercial-use) to the Pypath book to show you how does it work. I hope this helps, please let me know if you have any question.

**Comment by Tama-sha** [comment]
> About your other question, I checked the current state of license parameters in pypath, and I filled a few gaps to make it better usable in the database build process. Due to these recent developments, you'll need to update pypath from the git master branch (version should be `>=0.14.38`). I also [added a little section](https://pypath.omnipathdb.org/notebooks/manual.html#Example:-build-a-network-for-commercial-use) to the Pypath book to show you how does it work. I hope this helps, please let me know if you have any question.

![image](https://user-images.githubusercontent.com/106139199/224618852-e849f789-dc7c-4c12-8b8f-a850ec7732d8.png)
 Hi @deeenes thank you so much for your swift reply but seems like i caught myself in another issue after updating the pypath version. Now the first sample commands in the pypath book are not working. I am sorry but i am a novice

**Comment by Tama-sha** [comment]
> Hello @Tama-sha, The reason of this error is that your operating system's default character encoding is code page 1252, while the `resources.json` file distributed with `pypath` is UTF-8 encoded, as it is more or less a standard practice today. The solution is to run your Python interpreter with UTF-8 default encoding.

Hi @deeenes  i also tried to see the default encoding
I tried to find the default encoding using: 
import sys 
encoding = sys.getdefaultencoding() 
print(encoding) 
The output was:
utf-8

**Comment by deeenes** [comment]
I'll look into this soon, but just as a quick note, your screenshots are so small that I can not read them. You could instead just copy-paste the error and traceback as text, directly into your comments here. You can enclose it between triple backticks for proper formatting: `` ``` ...insert your multi-line code/output here... ``` ``.

About the encodings: apparently Python uses `cp1252` encoding when reading files, we have to find out why. As the built-in Python 3 `json` module should always use `UTF-8` as default, maybe Miniconda or some of your RC files modifies the default settings. The easiest way to reproduce the error is the following:

```
import json

path = '~\AppData\Local\Continuum\Miniconda3\lib/site-packages\pypath\resources\data\resources.json'
with open(path, 'r') as fp:
  resources = json.load(fp)
```

Then you can see if it fails, and you can check the encoding of fp:

```
fp
<_io.TextIOWrapper name='pypath/resources/data/resources.json' mode='r' encoding='UTF-8'>
```

Also, you could try to set the `LC_CTYPE` environment variable in PowerShell and then start Python and test the thing:

```
setx LC_CTYPE 'UTF-8'
python
```

Or set the variable in Cygwin Bash and start Python:

```
export LC_CTYPE='UTF-8'
python
```

However, we can also explicitly enforce the UTF-8 encoding in pypath, I will soon commit a patch for that. This will likely resolve your issue.

**Comment by Tama-sha** [comment]
Hi @deeenes 
I apologize for the screenshot quality.
The error i am getting after updating after updating the version to 0.14.38
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 3
      1 import locale
      2 locale.setlocale( locale.LC_ALL, 'en_US.UTF-8' )
----> 3 from pypath import resources
      4 rc = resources.get_controller()
      5 rc

File ~\AppData\Local\Continuum\miniconda3\lib\site-packages\pypath\resources\__init__.py:25
      1 #!/usr/bin/env python
      2 # -*- coding: utf-8 -*-
      3 
   (...)
     22 #  Website: http://pypath.omnipathdb.org/
     23 #
---> 25 from . import controller as _controller_mod
     28 def get_controller(reload = False):
     29     """
     30     Returns the resource controller.
     31 
     32     The controller instantiated only once by default, at the first access
     33     attempt. The instance is stored in the module and provided on demand.
     34     """

File ~\AppData\Local\Continuum\miniconda3\lib\site-packages\pypath\resources\controller.py:42
     39 import itertools
     41 import pypath.share.session as session_mod
---> 42 import pypath.share.common as common
     43 import pypath.internals.resource as resource_base
     44 import pypath.resources.network as netres

File ~\AppData\Local\Continuum\miniconda3\lib\site-packages\pypath\share\common.py:2643
   2635     """
   2636     Nest multiple functions into a single function.
   2637     """
   2639     return lambda x: functools.reduce(lambda x, f: f(x), (x,) + funcs)
   2642 def compr(
-> 2643         obj: Iterable | dict,
   2644         apply: Callable | None = None,
   2645         filter: Callable | None = None,
   2646     ) -> Iterable | dict:
   2647     """
   2648     Unified interface for list, dict, set and tuple comprehensions.
   2649 
   (...)
   2661         Same type as the input: a list, a dict or a set or a tuple.
   2662     """
   2664     _type = (
   2665         dict
   2666             if isinstance(obj, Mapping) else
   (...)
   2669         list
   2670     )

TypeError: unsupported operand type(s) for |: '_SpecialGenericAlias' and 'type'
```

**Comment by npalacioescat** [comment]
Hi @Tama-sha, the last error you're mentioning seems to be related to the type annotations which were introduced in Python 3.5 (2015), although the "|" operator came a bit later, not sure which specific version.
Which version of Python are you using? A more recent one will likely fix this specific issue.
